File: app/src/main/java/ch/bailu/aat/services/dem/tile/MultiCell4NE.java
package ch.bailu.aat.services.dem.tile;

public class MultiCell4NE extends MultiCell {
    /**
     *      a  b
     *      C  d
     */

    private short a, b, c, d;
    private int dzx, dzy;

    private final DemProvider demtile;
    private final int dim;
    private final int total_cellsize;


    public MultiCell4NE(final DemProvider dem) {
        demtile = dem;
        dim = dem.getDim().DIM;
        total_cellsize=Math.round(dem.getCellsize()*4f);
    }

    @Override
    public void set(int x) {
            _set(x);
        dzx=_delta_zx();
        dzy=_delta_zy();
    }

    private void _set(int x) {
        final int a=x-dim;
        final int b=a+1;
        final int c=x;
        final int d=x+1;
        this.a=demtile.getElevation(a);
        this.b=demtile.getElevation(b);
        this.c=demtile.getElevation(c);
        this.d=demtile.getElevation(d);
    }


    @Override
    public int delta_zx() {
        return dzx;
    }

    @Override
    public int delta_zy() {
        return dzy;
    }

    private int _delta_zx() {
        final int sum = ((b + d) - (a + c));
        return  (sum * 100) / total_cellsize;
    }

    private int _delta_zy() {
        final int sum = ((c + d) - (b + a));
        return (sum * 100)  / total_cellsize;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/elevation/tile/MultiCell4NE.kt
package ch.bailu.aat_lib.service.elevation.tile

class MultiCell4NE(private val demProvider: DemProvider) : MultiCell() {
    /**
     * a  b
     * C  d
     */
    private var a: Short = 0
    private var b: Short = 0
    private var c: Short = 0
    private var d: Short = 0
    private var dzx = 0
    private var dzy = 0

    private val dim = demProvider.getDimension().dimension
    private val totalCellDistance = Math.round(demProvider.getCellDistance() * 4f)

    override fun set(x: Int) {
        setCell(x)
        setDeltaZX()
        setDeltaZY()
    }

    private fun setCell(x: Int) {
        val a = x - dim
        val b = a + 1
        val c = x
        val d = x + 1
        this.a = demProvider.getElevation(a)
        this.b = demProvider.getElevation(b)
        this.c = demProvider.getElevation(c)
        this.d = demProvider.getElevation(d)
    }


    override fun deltaZX(): Int {
        return dzx
    }

    override fun deltaZY(): Int {
        return dzy
    }

    private fun setDeltaZX() {
        val sum = ((b + d) - (a + c))
        dzx =  (sum * 100) / totalCellDistance
    }

    private fun setDeltaZY() {
        val sum = ((c + d) - (b + a))
        dzy = (sum * 100) / totalCellDistance
    }
}
